src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Packaging.Licenses
+{
+    internal class LicenseExpressionTokenizer
+    {
+        private static Tuple<char, string> OpeningBracket = new Tuple<char, string>('(', "(");
+        private static Tuple<char, string> ClosingBracket = new Tuple<char, string>(')', ")");
+
+        private readonly string _value;
+        /// <summary>
+        /// A tokenizer for a license expression.
+        /// This implementation assumes that the input has been sanitized and that there are no invalid characters.
+        /// </summary>
+        /// <param name="value"></param>
+        internal LicenseExpressionTokenizer(string value)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+            _value = value.Trim();
+        }
+
+        /// <summary>
+        /// The valid characters for a license identifier are a-zA-Z0-9.-+
```

Comment:
>a-zA-Z0-9.-+ [](start = 62, length = 12)

Why didn't you use this regex expression?

Reply:
Perf concern. No need to load extra classes unless we truly need it. 

This tokenizer is very dumb and I'm changing a bit about it right now. 
It was something whipped up in 1 hour to just unblock myself :)

Reply:
My guess is that regex will be faster than a naive tokenizer. This may be worth a mini benchmark 😄

Reply:
It's actually rather difficult to write a regex to parse out the tokens for a valid license expression. 
It also destroys the debug-ability. 

Using regex to find invalid characters is a bit of an overkill, and we can't really do better than `O(n)` where n is the char count.

Reply:
To clarify, I'm only suggesting regex to simplify `HasValidCharacters` - not the rest of the tokenizer.

Reply:
Ahem, clear. 
Then the 2nd part applies :) 

regex's best time complexity `O(n)` + space for the construction.

Reply:
Regex engines have a few tricks up their sleeves (clever algorithms, SIMD, etc... See [this article by ripgrep](https://blog.burntsushi.net/ripgrep/#literal-optimizations)). If performance is the only concern, I think regex may be a viable alternative.

I suggested regex to simplify the code though. The performance benefits (if any) would just be an added bonus.

Reply:
Sounds good. 
Not too set on it either way. 
Thoughts @NuGet/nuget-client

Reply:
I don't care too much either way, I can change it, just don't see the point of jitting regex classes for a very dumb check.